### ISSUE TITLE
Design For Multithreaded Rendering

### ISSUE THREAD

**Issue by tomspilman** [mono/MonoGame#2772]
Looking forward past the XNA4 API we need to start considering how multithreaded render contexts fit into the current MonoGame rendering API.

I'm thinking the core concept is a `GraphicsContext`.  Which is where we move all the core render state and commands:

```
public class GraphicsContext : IDisposable
{
   public RasterizerState RasterizerState;
   public BlendState BlendState;
   public DepthStencilState DepthStencilState;
   public Rectangle ScissorRectangle;
   public Viewport Viewport;

   public void SetRenderTarget(...);
   public void SetVertexBuffer(...);
   public void SetIndexBuffer(...);

   internal Shader VertexShader;
   internal Shader PixelShader;

   public void Clear(...);

   public void DrawPrimitives(...);
   public void DrawIndexedPrimitives(...);
   public void DrawUserPrimitives(...);        
}
```

Then we change `GraphicsDevice` to only have the per-device stuff and inherit from `GraphicsContext`...

```
    public class GraphicsDevice : GraphicsContext
    {
       public DisplayMode DisplayMode;
       public PresentationParameters PresentationParameters;
       public GraphicsProfile GraphicsProfile;

       // Creates a deferred context for building command lists in parallel.
       public GraphicsContext CreateContext();

       // Submit commands from another context to be rendered by the GPU.
       public void DrawContext(GraphicsContext context);

       public void Present();
    }
```

This way to most users the MonoGame drawing API remains pretty identical to XNA4 except for this extra `GraphicsContext` base class in there.

Next is what to do with all the `GraphicsResource` classes like `VertexBuffer`, `Texture2D`, `BlendState`, etc.  These all hold on to a `GraphicsDevice` internally for various reasons.

In the cases where they need to create resources on the "GPU" they should be using `GraphicsDevice` and this path either is or should be fixed to be thread safe.

But in the cases of actual rendering they need to know what `GraphicsContext` to apply themselves to.  Luckily most of the resources are applied to the device/context like so:

```
deviceOrContext.BlendState = myBlendState;
deviceOrContext.SetVertexBuffer(myVertexBuffer);
deviceOrContext.Textures[0] = myTexture;
```

We just need to fix the internal APIs to pass both the `GraphicsDevice` (for creating resources) and `GraphicsContext` (for state/rendering) ...  which in most cases could be the same object.

The bigger problem would be `Effect` which just does this:

```
myEffect.CurrentTechnique.Passes[0].Apply();
```

It internally sets the shaders, constant buffers, textures, samplers, and render states on to the `GraphicsDevice` it was passed at construction.  To fix this we will need a new overloaded method in `EffectPass`:

```
public void Apply()
{
   Apply(GraphicsDevice);
}

public void Apply(GraphicsContext context)
{
    context.VertextShader = _vertexShader;
    context.PixelShader = _pixelShader;
    context.RasterizerState = _rasterizerState;
    context.BlendState = _blendState;
    context.DepthStencilState = _depthStencilState;

    // etc!
}
```
This then would require anyone using the multithreaded rendering features to be sure to call `EffectPass.Apply(GraphicsContext)`.

Thoughts?

**Comment by nkast** [comment]
The first step would be to move everything related to `ID3D12GraphicsCommandList` & `D3D12_GRAPHICS_PIPELINE_STATE_DESC` into  `GraphicsContext`. (GraphicsContext.DX11.cs?)

Then upgrade all DX11 objects to DX12 in `Device` class and pass a DX11 device to the `GraphicsContext` created by `D3D11On12CreateDevice`.

Once we have that we can move to true DX12 piece by piece and adding Metal/Valkan support as well. 
The challenge would be to support backward XNA functionality as we add multithreading extensions.
So far I know that `Effect` is one of them because it couples Shaders & Textures. The other would be States. Probably the third step would be refactoring all states/shades into a `PipelineState` object.


The concept of Context is DX11, now that DX12 interface is public maybe we should use different names.
`GraphicsCommandList` instead of `GraphicsContext`, 
`ExecuteCommand(...)` instead of `DrawContext(...)`, etc
I hope Valkan/Metal use similar names/concepts.

**Comment by nkast** [comment]
A simpler approach would be to keep `GraphicsDevice` as it is and use a 11on12 device underneath.
Just add a `CreateGraphicsCommandList()` and  `ExecuteCommandList()` method.

We don't have to implement a `CommandList.PipelineState` on top of DX11 and all `GraphicsDevice.*State` on top of DX12.
Traditional MG code will not see any benefit from DX12 but I believe that's already the case if we try to re-implement the checks of DX11.

**Comment by tomspilman** [comment]
I plan to find some time to evaluate all three of these APIs and come up with a high level class diagram to discuss here.  Until someone does this we really can't proceed or assume anything.

**Comment by KonajuGames** [comment]
Is the Vulkan API available yet? Or is still hidden behind private beta?

**Comment by Shqrdx** [comment]
The DirectX12 API looks like is not be supported the systems below Windows 10.. Thats a problem... or not ?

**Comment by tomspilman** [comment]
> Thats a problem... or not ?

It isn't a problem... we will still have the DX11 path for that.

**Comment by tomspilman** [comment]
> Is the Vulkan API available yet? 

No public API yet.  We do know however that it is being created from parts of ATI's Mantle, so that might be the best thing to review right now.

**Comment by KonajuGames** [comment]
Mantle isn't public yet either, is it?

**Comment by tomspilman** [comment]
> Mantle isn't public yet either, is it?

It is...

http://www.amd.com/Documents/Mantle-Programming-Guide-and-API-Reference.pdf

**Comment by KonajuGames** [comment]
Still nothing about Vulkan, and that Mantle API document is the only thing we'll see publicly about Mantle since AMD stopped development on it.

NVIDIA released a [DX12 Do's And Don'ts](https://developer.nvidia.com/dx12-dos-and-donts) document that will grow in time as more things are learnt about the best way to use DX12.

**Comment by tomspilman** [comment]
Closing this and moving discussion to https://github.com/mono/MonoGame/issues/4571.
